### PR TITLE
Support environment-based tox factors

### PR DESCRIPTION
**Pull request by rpkilby**
Taking a stab at #4.

### CODE REVIEW

**Comment by ryanhiebert** [comment]
This is good work, thank you.

Let me share with you a couple things that I am thinking on this front, so that I can get your thoughts on how you see these playing together.

I'm thinking that for factors that Travis supports in it's configuration, such as `python` and `os`, I'm wanting to use those names, and the values of the associated variables that Travis tells us about those configurations, in our configuration.

That doesn't preclude this possibility, but it may signal that instead of having the keys be the environment variable names, perhaps we need to have it be `env` instead, and have that be a two-level-deep mapping of names to values to env patterns.

So for instance this:

```ini
[tox:travis]
DJANGO =
    1.9: django19
    1.10: django110
```

You'd instead have

```ini
[tox:travis]
env =
    DJANGO:
        1.9: django19
        1.10: django110
```

Do you think that'd be reasonable, or do you think it'd be too ugly?

**Comment by rpkilby** [comment]
> I'm thinking that for factors that Travis supports in it's configuration, such as `python` and `os`, I'm wanting to use those names, and the values of the associated variables that Travis tells us about those configurations, in our configuration.

Happy to provide my thoughts, but I'm not really understanding the use case. How/why would you use this? 

Some thoughts:
- From a use case perspective, how you would integrate the `os` var? Tox controls python environments. How would you target different operating systems?
- What would the usage look like? Could you provide example travis & tox configs?
- The `python` and `os` travis factors have corresponding environment variables. Specifically `TRAVIS_PYTHON_VERSION` and `TRAVIS_OS_NAME`. Again, not understanding the use case, but what is insufficient with the following?

```ini
[tox:travis]
TRAVIS_OS_NAME =
    osx: ...
```

> instead of having the keys be the environment variable names, perhaps we need to have it be `env` instead, and have that be a two-level-deep mapping of names to values to env patterns.

From an API standpoint. I like the simplicity of just environment variable names. However, if it's necessary to support multiple factor... providers (?)... it might make sense to use separate sections. eg, 


```ini
[travis:env]
DJANGO =
    1.9: django19
    1.10: django110

[travis:os]
osx = ...
```

Side note, not sure if there is reason for prepending the sections w/ `tox:`. If so, could use `[tox:travis:env]` instead.

**Comment by ryanhiebert** [comment]
> How/why would you use this?

Here's the driving use-case. You want to test your app on macOS. Unfortunately, when you run under `osx` as the `os`, you have to make sure to set the `language` to `generic`. That is, not `python`, and therefore without any version selectors. It's likely at that point you'd like to only run a subset of tests on macOS, to avoid very long run times, etc.

Dealing with testing multiple OSs in Travis is complicated, and there's really nothing we can do about it. But we can provide easy hooks to make it simple to work with. See #19 and #25 for a bit more discussion of this problem.

> From a use case perspective, how you would integrate the os var? Tox controls python environments. How would you target different operating systems?

The `os` is exposed as `TRAVIS_OS_NAME`, in the same way that the `python` version is exposed as `TRAVIS_PYTHON_VERSION`.

> What would the usage look like? Could you provide example travis & tox configs?

I showed the idea I had for the `tox.ini` earlier. The `.travis.yml` would be more complicated, necessarily by the nature of the problem. I think there's some discussion of what it takes on #25.

> The python and os travis factors have corresponding environment variables. Specifically `TRAVIS_PYTHON_VERSION` and `TRAVIS_OS_NAME`. Again, not understanding the use case, but what is insufficient with the following?

My feeling now is that just using the actual environment variable names is probably not a good idea for Travis configuration items that we know how to map anyway. It's just too cryptic a name, and though it's documented, it's a fairly obscure feature of Travis. Using the same keys as the travis config will make things much more user-friendly.

You had an excellent idea with the `travis:env` section. I really like it. It's obvious and clear. I _don't_ want to have similar sections for `os`, `language`, etc, but I really like it for `env`. It also separates the namespace of the env from the other name, avoiding backward compatibility issues pretty nicely.

AFA `travis:env` vs `tox:travis:env`, I chose `tox:travis` to match `tox:jenkins`, which, IIUC, tox supports natively. However, I think that I like `travis:env` better for this new section.

**Comment by rpkilby** [comment]
It seems that the API is shaping up to something along the lines of:

```ini
[tox:travis]  # maybe rename to [travis]?
python =
    2.7: py27
    3.5: py35

os =
    osx: py{27,34}-{spam}

[travis:env]
DJANGO =
    1.9: django19
    1.10: django110
```

I like this approach. First-class support can be provided for some travis factors, while the rest can be handled with env vars.

Is there anything to do beyond using the separate env section?

**Comment by ryanhiebert** [comment]
Nope, I think that'd good enough to merge your work. Thank you for your feedback and your work. I agree that the API seems to be shaping up nicely.

I have some ideas about what we might end up doing with the bare `travis` section, but there's no need to figure that out for this PR.

**Comment by rpkilby** [comment]
So the PR grew a bit.

From the above example, I realized that the factors needed comma/brace support (`parse_dict`). eg, 
```ini
[travis]
os =
    osx: py{27,34}, docs
```

However, this in turn implies that you can get multiple envs from the factors (`reduce_factors`). After separating out the `[travis:env]` section, it was pretty straightforward to add the `[travis]` section separate from original `[tox:travis]`.

**Comment by ryanhiebert** [comment]
You did grow this PR ;-) That's good and bad. It's awesome, because it means this project just gained some awesome momentum. It's less awesome for you because it'll end up taking a bit more to get this PR in, because it includes more design work that we'll need to vet before we merge. I hope you don't mind. I definitely appreciate everything you've put into this, and if you decide to step away from the struggle getting this PR in, I'll end up using your work as a starting point for completing it. Thank you!

You've got here that it will default to the old behavior if the `tox:travis` exists. This seems reasonable to me, as long as we warn when there is a `tox:travis` section along with _either_ of the new sections, `travis` or `travis:env`.

The main design decision that I'm questioning on this is when to use, or not use, the "autoenv" feature. These are the magic defaults that were the original reason that I made tox-travis. In this patch, it looks like you've chosen to make the answer "when the current os or python version have something specified in the `travis` section." I'm not sure that's the right answer, because the autoenv defaults only really deal with the Python version. The `os` really has no bearing on it.

I think the right answer is to use the autoenv default when the `TRAVIS_PYTHON_VERSION` is set, but the value has not been specified in the `python` key of the `travis` section. This is actually different from our current behavior, which uses autoenv also if that variable is unset, but I'm OK with that backward incompatibility. We'll document it in the release notes.

Anything except the `python` key, which has to have special handling to deal with the autoenv feature, can just be used just as if it were a key in the `travis:env` section, using a simple mapping:

```python
{
  'os': 'TRAVIS_OS_NAME',
  'language': 'TRAVIS_LANGUAGE',  # undocumented?
  'python': 'TRAVIS_PYTHON_VERSION',
}
```

File: src/tox_travis.py [hunk 1]
```diff
@@ -99,13 +101,104 @@ def get_default_envlist(version):
...
+def reduce_factors(factors):
+    """Reduce sets of factors into a single list of envs.
+
+        >>> reduce_factors([['py27', 'py35'], ['django110']])
+        ['py27-django110', 'py35-django110']
+
+    """
+    envs = product(*factors)
+    envs = ['-'.join(env) for env in envs]
+    return envs
```
Comment (ryanhiebert) [inline src/tox_travis.py]: Before I forget about it, duplicated factors need to be collapsed. For instance:

```pycon
>>> reduce_factors(['py27', 'py35'], ['py35'])
['py27-py35', 'py35']
```

The first is obviously non-sensical, But `py35-py35` would be redundant, but also clear that it's likely wanting to just run the `py35` environment.

**Comment by rpkilby** [comment]
> You've got here that it will default to the old behavior if the tox:travis exists. This seems reasonable to me, as long as we warn when there is a tox:travis section along with either of the new sections, travis or travis:env.

Assertion checks added. Assertions seem to make sense (as opposed to warnings), since this would otherwise be invalid behavior. 

----

> I think the right answer is to use the autoenv default when the `TRAVIS_PYTHON_VERSION` is set, but the value has not been specified in the python key of the travis section. This is actually different from our current behavior, which uses autoenv also if that variable is unset, but I'm OK with that backward incompatibility. We'll document it in the release notes.

That makes sense. In practice, this would mean `python`/autoenv is mutually exclusive to `os`. You won't run into weird combinations of `py35-py35`.

> ```py
> >>> reduce_factors(['py27', 'py35'], ['py35'])
> ['py27-py35', 'py35']
> ```

> The first is obviously non-sensical, But `py35-py35` would be redundant, but also clear that it's likely wanting to just run the `py35` environment.

Two thoughts:
- The above is nonsensical, and given the above changes shouldn't happen. If it does happen...
- It's not clear how this could reliably and meaningfully be de-duplicated. What should happen for `

```py
 >>> reduce_factors(['py27', 'py35'], ['py35', 'docs'])
```

or 

```py
 >>> reduce_factors(['py27', 'py35'], ['py34', 'docs'])
```

**Comment by ryanhiebert** [comment]
Rather than the contrived example, let me present this more realistic example:

```ini
[tox]
envlist = py{27,34,35}, docs

[travis]
os = 
    linux: py{27,35}, docs
    osx: py{27,35}
python =
    py35: py35, docs
```

This configuration would mean that on Linux we want to generate the docs, but we don't care if the docs generate on OSX, so we won't test that. The docs env should run in the same test run as Python 3.5 on Linux, but be ignored for OSX.

The easy problem is figuring out which envs to run. If the envlist in the `tox` section were `py{27,34,35}, docs`, it's clear that it should run `py35` and `docs`. If we match by factors, that's easy enough, and having the cross-list expand to including non-sensical stuff like `py27-py35` and `docs-docs` won't be a problem: they'll still match appropriately based on those factors.

The challenge comes in the cases where there is no matching env declared in the tox config already, either in the `envlist` or in any of the `testenv:` sections. Currently we only support a single factor, so it's reasonable for us to say "if nothing matches, just use the list given verbatim". That means that for the very simple case, the `tox` section doesn't even need to define the `envlist`, because it'll just use the `py35` or whatever env by default.

Moving to multiple configurations that need to be joined makes that difficult and confusing. Using the output of `reduce_factors` seems like a bad idea, because it'll be so _easy_ for it to produce complete jibberish env names.

It's a useful feature in simple cases, but it breaks down in the face of multiple configurations. Right now, I'm thinking that the rule for that feature probably needs to be that it only works if there is only one configuration specified, and that configuration needs to be `python` (since the autoenv feature will specify `python` automatically, making there be implicitly _two_ configurations if any other is set). If we ever make autoenv optional, then we can reasonably allow this feature to be used by another configuration instead of just `python`.

------

In summary, I no longer think we need to "simplify" the envlist as produced by `reduce_factors`, it's probably fine. However, we _do_ need to make sure that the passthrough envlist feature only works in cases where a single configuration (python, env var, etc) is specified in a given combination. That is, when only _one_ list is provided to `reduce_factors`.

**Comment by ryanhiebert** [comment]
For the warning, I think I'd like it to be a bit different than you've created it, and also different from how I previously described it. Thank you for the concrete implementation, it made it much easier for me to think about it.

* It should be a logged warning, and not raise an exception.
* Having both `tox:travis` and another `travis` section should be fine, it should run without any issues, apart from the deprecation warning.
* For backward compatibility, transform the `tox:travis` section internally, and treat it as if it were the `python` key of the `travis` section.
* If both `tox:travis` and the `python` key of `travis` exist, just treat them as multiple factors.
* Display the deprecation whenever the `tox:travis` section exists, even if no other travis sections exist. It should probably be printed on stderr.

I'd like the warning message to be something along the lines of:

> The [tox:travis] section is deprecated in favor of the "python" key in the [travis] section.


Feel free to nit-pick that message, but I'd like it to

1. Let us know that the old section is _deprecated_.
2. Let us know what the preferred alternative is.

While the `tox:travis` section will be deprecated, and its usage removed from the documentation, I'm not inclined to be setting a timeframe for removing the feature. It's easy enough for it to keep working, and I'm inclined to avoid breaking things unless there's a good reason.

-----

Also, and not terribly important given that we're not going to raise an error anyway, but unfortunately `assert` is not useful for making runtime assertions. I know, it sounds silly. It's more of a safety net for making sure that things are working as you expect they are when you're testing in development. Because `assert` statements are _removed_ from the compiled python when run in optimized mode (`python -O`, typically for a production environment, though I usually don't bother), they really can't be relied on for checks in production.

-----

Also, I'll re-iterate that if you get tired of the back and forth with me, it's perfectly fine to leave me to just finish merging this myself. There's a ton of good work here, and I promise it won't be wasted. Just let me know if you get to that place.

**Comment by rpkilby** [comment]
Hi @ryanhiebert - I'm happy to continue working on this, but am short on bandwidth for a couple days. If you know what needs to be done and want to go ahead and knock this out, then by all means don't wait on me.
